File: apps/api/src/modules/auth/auth.service.ts
import { AccountService } from '../account/account.service';
import { Injectable, UnauthorizedException } from '@nestjs/common';
import { RegisterUserDto } from './dto/register-user.dto';
import * as bcrypt from 'bcrypt';
import { JwtService } from '@nestjs/jwt';
import { JwtPayload } from './interfaces/jwt-payload.interface';
import { AccountRole } from '../../shared/enums/account-role.enum';
import { AuthAccountResponse } from './interfaces/auth-account-response.interface';

@Injectable()
export class AuthService {
  constructor(
    private readonly accountService: AccountService,
    private readonly jwtService: JwtService,
  ) {}

  // validateUser là tên hàm chuẩn của Passport Local Strategy
  // Hàm này sẽ được gọi khi người dùng đăng nhập
  // Không thay đổi tên hàm này để Passport có thể nhận diện
  async validateUser(
    email: string,
    pass: string,
  ): Promise<{
    access_token: string;
    account: AuthAccountResponse;
  }> {
    // 1. Tìm tài khoản theo email
    const account = await this.accountService.findByEmail(email);

    // 2. Kiểm tra tài khoản có tồn tại không và mật khẩu có khớp không
    if (!account) {
      throw new UnauthorizedException('Invalid credentials!');
    }

    // So sánh mật khẩu đã hash bằng bcrypt với mật khẩu người dùng nhập vào
    const isMatch = await bcrypt.compare(pass, account.password);
    if (!isMatch) {
      throw new UnauthorizedException('Invalid credentials!');
    }

    // 3. Tạo JWT token
    // Tạo payload cho JWT, có thể thêm thông tin khác nếu cần
    const payload: JwtPayload = {
      sub: account.id,
      email: account.email,
      role: account.role,
      name: account.name,
    };

    return {
      // Sử dụng JwtService để tạo token từ payload
      // access_token là tên chuẩn, có thể đổi nếu cần
      access_token: await this.jwtService.signAsync(payload),
      // Tra trả thông tin tài khoản đã loại bỏ mật khẩu
      account: {
        id: account.id,
        email: account.email,
        role: account.role,
        name: account.name,
        avatar_url: account.avatar_url,
      },
    };
  }

  async register(dto: RegisterUserDto) {
    // Let AccountService.create() handle all validation and processing
    // (email normalization, existence checking, password hashing)
    await this.accountService.create({
      ...dto,
      role: AccountRole.USER, // Default role for registration
    });

    // Return success message without exposing user data
    return {
      message: 'Registration successful! Please login to continue.',
    };
  }

  // Handle Google login: create user if not exists, then generate JWT
  async handleGoogleLogin(googleProfile: {
    googleId: string;
    email: string;
    name: string;
    avatar?: string;
    provider: string;
    emailVerified?: boolean;
  }): Promise<{
    access_token: string;
    account: AuthAccountResponse;
  }> {
    // 1. Check if user exists by email
    let account = await this.accountService.findByEmail(googleProfile.email);

    // 2. If user doesn't exist, create a new one
    if (!account) {
      // Create a new account for Google user
      // Generate a random password for Google users (they won't use it)
      const randomPassword =
        Math.random().toString(36).slice(-12) +
        Math.random().toString(36).slice(-12);

      account = await this.accountService.create({
        email: googleProfile.email,
        name: googleProfile.name,
        password: randomPassword, // Google users won't use this password
        role: AccountRole.USER, // Default role for Google users
      });

      // Update avatar if provided
      if (googleProfile.avatar && account.id) {
        await this.accountService.update(account.id, {
          avatar_url: googleProfile.avatar,
        });
        account.avatar_url = googleProfile.avatar;
      }
    }

    // 3. Generate JWT token with the database user ID
    const payload: JwtPayload = {
      sub: account.id, // Use database user ID, not googleId
      email: account.email,
      role: account.role,
      name: account.name,
      avatar: account.avatar_url,
      provider: googleProfile.provider,
    };

    return {
      access_token: this.jwtService.sign(payload),
      account: {
        id: account.id,
        email: account.email,
        role: account.role,
        name: account.name,
        avatar_url: account.avatar_url,
      },
    };
  }

  // Tạo JWT từ GoogleProfile hoặc Account
  generateJwt(payload: JwtPayload) {
    // Có thể tuỳ chỉnh payload cho phù hợp
    return this.jwtService.sign(payload);
  }
}

File: apps/api_go/internal/modules/auth/service/auth_service.go
package service

import (
	"errors"
	"math/rand"
	"strings"
	"time"

	"github.com/golang-jwt/jwt/v5"
	"golang.org/x/crypto/bcrypt"

	"api_go/internal/config"
	"api_go/internal/domain"
)

type authService struct {
	config      *config.Config
	accountSvc  domain.AccountService
	accountRepo domain.AccountRepository
}

// NewAuthService creates a new AuthService instance
func NewAuthService(cfg *config.Config, accountSvc domain.AccountService, accountRepo domain.AccountRepository) domain.AuthService {
	return &authService{
		config:      cfg,
		accountSvc:  accountSvc,
		accountRepo: accountRepo,
	}
}

// ValidateUser validates email/password and returns JWT + account info
func (s *authService) ValidateUser(email, password string) (*domain.AuthResponseDTO, error) {
	// 1. Normalize email
	normalizedEmail := strings.ToLower(strings.TrimSpace(email))

	// 2. Find account by email (need raw entity with password)
	account, err := s.accountRepo.FindByEmail(normalizedEmail)
	if err != nil {
		return nil, err
	}
	if account == nil {
		return nil, errors.New("invalid credentials")
	}

	// 3. Compare password
	if err := bcrypt.CompareHashAndPassword([]byte(account.Password), []byte(password)); err != nil {
		return nil, errors.New("invalid credentials")
	}

	// 4. Generate JWT token
	payload := domain.JWTPayload{
		Sub:      account.ID,
		Email:    account.Email,
		Role:     account.Role,
		Name:     account.Name,
		Provider: "local",
	}
	if account.AvatarURL != nil {
		payload.Avatar = *account.AvatarURL
	}

	token, err := s.GenerateJWT(payload)
	if err != nil {
		return nil, err
	}

	// 5. Build response
	return &domain.AuthResponseDTO{
		AccessToken: token,
		Account: domain.AuthAccountResponse{
			ID:        account.ID,
			Email:     account.Email,
			Role:      account.Role,
			Name:      account.Name,
			AvatarURL: account.AvatarURL,
		},
	}, nil
}

// Register creates a new user account
func (s *authService) Register(dto domain.RegisterDTO) error {
	_, err := s.accountSvc.Create(domain.CreateAccountDTO{
		Email:    dto.Email,
		Name:     dto.Name,
		Password: dto.Password,
		Role:     string(domain.AccountRoleUser), // Default role for registration
	})
	return err
}

// HandleGoogleLogin handles Google OAuth login (creates user if not exists)
func (s *authService) HandleGoogleLogin(profile domain.GoogleProfile) (*domain.AuthResponseDTO, error) {
	// 1. Check if user exists by email
	account, err := s.accountRepo.FindByEmail(profile.Email)
	if err != nil {
		return nil, err
	}

	// 2. If user doesn't exist, create a new one
	if account == nil {
		// Generate a random password for Google users (they won't use it)
		randomPassword := generateRandomPassword(24)

		// Create account via service (handles password hashing)
		created, err := s.accountSvc.Create(domain.CreateAccountDTO{
			Email:    profile.Email,
			Name:     profile.Name,
			Password: randomPassword,
			Role:     string(domain.AccountRoleUser),
		})
		if err != nil {
			return nil, err
		}

		// Update avatar if provided
		if profile.Avatar != "" && created.ID != 0 {
			_, _ = s.accountSvc.Update(created.ID, domain.UpdateAccountDTO{
				AvatarURL: &profile.Avatar,
			})
		}

		// Fetch the created account
		account, err = s.accountRepo.FindByEmail(profile.Email)
		if err != nil {
			return nil, err
		}
	}

	// 3. Generate JWT token
	payload := domain.JWTPayload{
		Sub:      account.ID,
		Email:    account.Email,
		Role:     account.Role,
		Name:     account.Name,
		Provider: "google",
	}
	if account.AvatarURL != nil {
		payload.Avatar = *account.AvatarURL
	}

	token, err := s.GenerateJWT(payload)
	if err != nil {
		return nil, err
	}

	return &domain.AuthResponseDTO{
		AccessToken: token,
		Account: domain.AuthAccountResponse{
			ID:        account.ID,
			Email:     account.Email,
			Role:      account.Role,
			Name:      account.Name,
			AvatarURL: account.AvatarURL,
		},
	}, nil
}

// GenerateJWT generates a JWT token from payload
func (s *authService) GenerateJWT(payload domain.JWTPayload) (string, error) {
	claims := jwt.MapClaims{
		"sub":      payload.Sub,
		"email":    payload.Email,
		"role":     payload.Role,
		"name":     payload.Name,
		"avatar":   payload.Avatar,
		"provider": payload.Provider,
		"iat":      time.Now().Unix(),
		"exp":      time.Now().Add(7 * 24 * time.Hour).Unix(), // 7 days
	}

	token := jwt.NewWithClaims(jwt.SigningMethodHS256, claims)
	return token.SignedString([]byte(s.config.JWTSecret))
}

// ValidateJWT validates a JWT token and returns the payload
func (s *authService) ValidateJWT(tokenString string) (*domain.JWTPayload, error) {
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		// Validate signing method
		if _, ok := token.Method.(*jwt.SigningMethodHMAC); !ok {
			return nil, errors.New("unexpected signing method")
		}
		return []byte(s.config.JWTSecret), nil
	})

	if err != nil {
		return nil, err
	}

	if claims, ok := token.Claims.(jwt.MapClaims); ok && token.Valid {
		payload := &domain.JWTPayload{
			Email:    claims["email"].(string),
			Role:     claims["role"].(string),
			Name:     claims["name"].(string),
			Provider: claims["provider"].(string),
		}

		// Handle sub as float64 (JSON numbers are decoded as float64)
		if sub, ok := claims["sub"].(float64); ok {
			payload.Sub = uint(sub)
		}

		if avatar, ok := claims["avatar"].(string); ok {
			payload.Avatar = avatar
		}

		return payload, nil
	}

	return nil, errors.New("invalid token")
}

// GetUserFromToken extracts user info from JWT token
func (s *authService) GetUserFromToken(tokenString string) (*domain.AccountResponseDTO, error) {
	payload, err := s.ValidateJWT(tokenString)
	if err != nil {
		return nil, err
	}

	// Fetch user from database
	account, err := s.accountRepo.FindByEmail(payload.Email)
	if err != nil {
		return nil, err
	}
	if account == nil {
		return nil, errors.New("user not found")
	}

	return &domain.AccountResponseDTO{
		ID:        account.ID,
		Email:     account.Email,
		Name:      account.Name,
		AvatarURL: account.AvatarURL,
	}, nil
}

// generateRandomPassword generates a random password
func generateRandomPassword(length int) string {
	const charset = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789!@#$%^&*"
	seededRand := rand.New(rand.NewSource(time.Now().UnixNano()))
	b := make([]byte, length)
	for i := range b {
		b[i] = charset[seededRand.Intn(len(charset))]
	}
	return string(b)
}
